### PR TITLE
V3 literate mode add quote thing at long last

### PR DESCRIPTION
**Pull request by lyxal**

### CODE REVIEW

**Comment by Lyxal** [comment]
I will obviously need to add dedicated tests for this

**Comment by ysthakur** [comment]
I don't envy you for having to wrangle with the parser. I'll need to look at this later, but can we go back to having both lexers extend a common trait, and have that trait accept a type parameter T (for tge two different Token/LitToken classes)?

**Comment by ysthakur** [comment]
Actually, I don't remember if there were any actual benefits to havkng a common trait, so never mind

File: shared/src/vyxal/parsing/Lexer.scala [hunk 1]
```diff
@@ -150,8 +160,88 @@ object Lexer:
...
+          merged.last.copy(value =
+            merged.last.value.asInstanceOf[String] +
+              token.value.asInstanceOf[String]
+          )
+        else merged += token
+      else merged += token
+
+    // Now, bind the move right tokens to the next token
+
+    var bound = ListBuffer[LitToken | Tuple2[LitToken, Int]]()
```
Comment (ysthakur) [inline shared/src/vyxal/parsing/Lexer.scala:185]:
```suggestion
    var bound = ListBuffer[LitToken | (LitToken, Int)]()
```
You used Copilot again, didn't you?

File: shared/src/vyxal/parsing/Lexer.scala [hunk 2]
```diff
@@ -150,8 +160,88 @@ object Lexer:
...
+          case last: LitToken =>
+            if last.tokenType == TokenType.MoveRight then
+              bound.dropRightInPlace(1)
+              bound += (token -> last.value.asInstanceOf[String].length)
+            else bound += token
+      else bound += token
+
+    // Move the tuple2's to the right, storing indices of where they were
+
+    while bound.exists(_.isInstanceOf[Tuple2[LitToken, Int]]) do
```
Comment (ysthakur) [inline shared/src/vyxal/parsing/Lexer.scala:199]:
```suggestion
    while bound.exists(_.isInstanceOf[(LitToken, Int)]) do
```

File: shared/src/vyxal/parsing/Lexer.scala [hunk 3]
```diff
@@ -150,8 +160,88 @@ object Lexer:
...
+        else merged += token
+      else merged += token
+
+    // Now, bind the move right tokens to the next token
+
+    var bound = ListBuffer[LitToken | Tuple2[LitToken, Int]]()
+    for token <- merged do
+      if bound.nonEmpty then
+        bound.last match
+          case _: Tuple2[LitToken, Int] => bound += token
```
Comment (ysthakur) [inline shared/src/vyxal/parsing/Lexer.scala:189]:
```suggestion
          case _: (LitToken, Int) => bound += token
```

File: shared/src/vyxal/parsing/Lexer.scala [hunk 4]
```diff
@@ -150,8 +160,88 @@ object Lexer:
...
+              bound.dropRightInPlace(1)
+              bound += (token -> last.value.asInstanceOf[String].length)
+            else bound += token
+      else bound += token
+
+    // Move the tuple2's to the right, storing indices of where they were
+
+    while bound.exists(_.isInstanceOf[Tuple2[LitToken, Int]]) do
+      val index = bound.indexWhere(_.isInstanceOf[Tuple2[LitToken, Int]])
+      val (token, offset) = bound(index).asInstanceOf[Tuple2[LitToken, Int]]
```
Comment (ysthakur) [inline shared/src/vyxal/parsing/Lexer.scala:201]: .

File: shared/src/vyxal/parsing/Lexer.scala (hunk 1 above)
Comment (Lyxal) [reply shared/src/vyxal/parsing/Lexer.scala:185]: Nope, this was me. I liked the explicit tuple2

**Pull request merged**